File: 2962_Count_Subarrays_Where_Max_ElementAppearsAtLeastKTimes.py
from typing import List

class Solution:
    def countSubarrays(self, nums: List[int], k: int) -> int:
        m = max(nums)

        l =len(nums)
        indexes =[-1]
        for i in range(l):
            if nums[i]==m:
                indexes.append(i)

        l_i =len(indexes)
        if l_i<1+k:
            return 0
        
        ans = 0
        for i in range(1,l_i-k+1):
            ans = ans + ((indexes[i]-indexes[i-1]) * (l-indexes[i+k-1]))

        return ans


print(Solution().countSubarrays([1,3,2,3,3],2))

print(Solution().countSubarrays([1,3,2,1,2],2))

# [1,3,2,3,3], k = 2
# -1 [1,3,4] 5

# 2*2 + 2*1 = 6

File: 2962_CountSubarraysWhereMaxElementAppearsAtLeastKTimes.cpp
#include<bits/stdc++.h>
using namespace std;

class Solution {
public:
    long long countSubarrays(vector<int>& nums, int k) {
        long long count = 0;
        int n=nums.size(), maxNum = *max_element(nums.begin(), nums.end()), maxCount = 0, left = 0;
        
        for (int right=0; right<n; ++right){
            if (nums[right]==maxNum) ++maxCount;

            while (maxCount>=k){
                count+=n-right;
                if (nums[left]==maxNum) --maxCount;
                ++left;
            }
        }

        return count;
    }
};